### PR TITLE
[#2002] Add explicit return type annotations to TypeScript functions in *.vue files

### LINKED ISSUE #2002: Explicit return type annotation for TypeScript functions
### What feature(s) would you like to see in RepoSense
Specify the return type of relevant functions in the frontend codebase. This is important because it can help catch errors early in the development process. 

### Is the feature request related to a problem?
Currently, many functions on the frontend do not have a return type even after the files were migrated to TypeScript.  If we don't specify the return types for functions, TypeScript will try to infer it based on the function's implementation. However, this inference can sometimes be incorrect, leading to unexpected behavior or errors.

By specifying the return type explicitly, we can help TypeScript catch any type errors before the code even runs. Additionally, specifying return types can make the code more readable and self-documenting, as it makes it clear what type of value the function will return.

### If possible, describe the solution
For official TypeScript docs on using return types, one can refer to https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#return-type-annotations.

### Additional context
Discussed in #1969

### PR DESCRIPTION
**Pull request by supermii2**
Part of #2002 

```
Many frontend functions lack return types after migration to
TypeScript, which can cause errors due to TypeScript's attempted type
inference being incorrect. Specifying return type explicitly can allow
TypeScript to catch errors before the code is run.

Let us add return types annotations to TypeScript functions within Vue
files.
```

### CODE REVIEW

File: frontend/src/components/c-ramp.vue [hunk 1]
```diff
@@ -171,21 +171,21 @@ export default defineComponent({
...
       return (new Date(this.udate).valueOf() - new Date(date).valueOf()) / (total + window.DAY_IN_MS);
     },
 
     // get duration in miliseconds between 2 date
-    getTotalForPos(sinceDate: string, untilDate: string) {
+    getTotalForPos(sinceDate: string, untilDate: string): number {
       return new Date(untilDate).valueOf() - new Date(sinceDate).valueOf();
     },
-    getRampColor(commit: CommitResult, slice: Commit) {
+    getRampColor(commit: CommitResult, slice: Commit): number | '-deletes' {
```
Comment (ckcherry23) [inline frontend/src/components/c-ramp.vue:188]: We could add a return type for `getSliceColor()` on line 194 as well.

File: frontend/src/components/c-summary-charts.vue [hunk 2]
```diff
@@ -372,7 +372,13 @@ export default defineComponent({
...
     },
   },
-  data() {
+  data(): {
+    drags: Array<number>,
+      activeRepo: string | null,
+      activeUser: string | null,
+      activeTabType: string | null,
+      isTabOnMergedGroup: boolean,
+      } {
```
Comment (ckcherry23) [inline frontend/src/components/c-summary-charts.vue:381]: Can you fix the space formatting here?

**Review by ckcherry23 — COMMENTED** [review]
Great job on working on your first frontend PR! Have left some minor comments to address. I hope adding the types helped you get acquainted well with the frontend codebase!

File: frontend/src/components/c-summary-charts.vue (hunk 2 above)
Comment (ckcherry23) [inline frontend/src/components/c-summary-charts.vue:381]:
```suggestion
    } {
```

File: frontend/src/components/c-ramp.vue (hunk 1 above)
Comment (supermii2) [reply frontend/src/components/c-ramp.vue:188]: Fixed! Thank you

File: frontend/src/components/c-summary-charts.vue (hunk 2 above)
Comment (supermii2) [reply frontend/src/components/c-summary-charts.vue:381]: Gradle's linter prevents the build if the space formatting is changed so I'll keep it unchanged for now

File: frontend/src/app.vue [hunk 3]
```diff
@@ -38,7 +38,16 @@ const app = defineComponent({
...
-  data() {
+  data(): {
+    repos: { [key: string]: Repo },
+    users: Array<Repo>
+      userUpdated: boolean,
+      loadingOverlayOpacity: number,
+      tabType: string,
+      creationDate: string,
+      reportGenerationTime: string,
+      errorMessages: {[key: string]: ErrorMessage }
```
Comment (vvidday) [inline frontend/src/app.vue:49]:
```suggestion
      errorMessages: { [key: string]: ErrorMessage }
```

File: frontend/src/views/c-authorship.vue [hunk 4]
```diff
@@ -125,7 +125,23 @@ const filesSortDict = {
...
 };
 
-function authorshipInitialState() {
+function authorshipInitialState(): {
+    isLoaded: boolean,
+    selectedFiles: Array<AuthorshipFile>,
+    filterType: FilterType,
+    selectedFileTypes: Array<string>,
+    fileTypes: Array<string>
+    filesLinesObj: {[key: string]: number}
```
Comment (vvidday) [inline frontend/src/views/c-authorship.vue:134]:
```suggestion
    filesLinesObj: { [key: string]: number }
```

File: frontend/src/views/c-authorship.vue [hunk 5]
```diff
@@ -157,7 +173,23 @@ export default defineComponent({
...
     'deactivate-tab',
   ],
-  data() {
+  data(): {
+    isLoaded: boolean,
+    selectedFiles: Array<AuthorshipFile>,
+    filterType: FilterType,
+    selectedFileTypes: Array<string>,
+    fileTypes: Array<string>
+    filesLinesObj: {[key: string]: number}
```
Comment (vvidday) [inline frontend/src/views/c-authorship.vue:182]:
```suggestion
    filesLinesObj: { [key: string]: number }
```

Comment (vvidday) [reply frontend/src/views/c-authorship.vue:182]: Might be worth looking into implementing a linting rule for the spaces in such types... (in a separate PR)

**Review by vvidday — CHANGES_REQUESTED** [review]
Thanks for your work on this! I believe most of the raw `.ts` files are still lacking return types - it might be too much for one PR and I'm ok with having this PR focus on just the functions in `.vue` files! In that case, we might want to update the title of this PR to reflect that, and change "Fixes" to "Part of" to prevent this PR from closing the entire issue.

Also as Jonas mentioned, it might be worth looking into if there's an eslint rule that can help enforce the return types! I believe it's possible to only enable the rule for `.vue` files too, and we can proceed to enable it for general `.ts` files once that portion has been completed in a separate PR.

**Comment by ckcherry23** [comment]
> Also, I think the title should include the issue number in front!

@supermii2 Remember to add the issue number in the front of the title!

**Comment by supermii2** [comment]
Closed due to a sync with master, opening it again

**Comment by supermii2** [comment]
> Thanks for your work on this! I believe most of the raw `.ts` files are still lacking return types - it might be too much for one PR and I'm ok with having this PR focus on just the functions in `.vue` files! In that case, we might want to update the title of this PR to reflect that, and change "Fixes" to "Part of" to prevent this PR from closing the entire issue.
> 
> Also as Jonas mentioned, it might be worth looking into if there's an eslint rule that can help enforce the return types! I believe it's possible to only enable the rule for `.vue` files too, and we can proceed to enable it for general `.ts` files once that portion has been completed in a separate PR.

Suggestion added, thank you!

**Comment by supermii2** [comment]
>Also as Jonas mentioned, it might be worth looking into if there's an eslint rule that can help enforce the return types! I believe it's possible to only enable the rule for .vue files too, and we can proceed to enable it for general .ts files once that portion has been completed in a separate PR

Hi, I've had to change the linting rules for unused variables to use the one built for typescript because it was creating false positives. Thank you!

I'll revert the changes done to *.ts in the commit after this comment.

**Comment by vvidday** [comment]
> > Also as Jonas mentioned, it might be worth looking into if there's an eslint rule that can help enforce the return types! I believe it's possible to only enable the rule for .vue files too, and we can proceed to enable it for general .ts files once that portion has been completed in a separate PR
> 
> Hi, I've had to change the linting rules for unused variables to use the one built for typescript because it was creating false positives. Thank you!
> 
> I'll revert the changes done to *.ts in the commit after this comment.

Could you elaborate (maybe with some examples) regarding the false positives, and how changing the rule fixes it? Also, could you check on the grammar of the final sentence in the proposed commit message? Thanks!

File: frontend/src/main.ts [hunk 6]
```diff
@@ -8,7 +8,7 @@ import 'muicss/dist/css/mui.min.css';
 import 'normalize.css/normalize.css';
 import 'vue-loading-overlay/dist/vue-loading.css';
 import '@fontsource/titillium-web';
-
+//
```
Comment (vvidday) [inline frontend/src/main.ts:11]: Was this change intentional?

**Comment by MarcusTXK** [comment]
@supermii2 bump on this, I think this PR is almost complete, just need some small changes and we can merge this in

File: frontend/src/main.ts (hunk 6 above)
Comment (supermii2) [reply frontend/src/main.ts:11]: Sorry this change was not intentional!

**Comment by supermii2** [comment]
> Could you elaborate (maybe with some examples) regarding the false positives, and how changing the rule fixes it? Also, could you check on the grammar of the final sentence in the proposed commit message? Thanks!

```
    sortingFunction(): (a: Commit, b: Commit) => number {
      const commitSortFunction = this.commitsSortType === CommitsSortType.Time
        ? (commit: Commit): string => commit.date
        : (commit: Commit): number => commit.insertions;

      return (a: Commit, b: Commit): number => (this.toReverseSortedCommits ? -1 : 1)
        * window.comparator(commitSortFunction)(a, b);
    },
```

Here's a potential false positive found in c-zoom.vue. The default no-unused-vars marks a and b used in the return type as unused variables, which is fixed by using the typescript version.

**Review by vvidday — APPROVED** [review]
Hi @supermii2, I added the missing return types for the new file (`c-zoom-commit-message`) & updated the proposed commit message to move this PR along for now.

Please take note for future that whenever you finish the requested changes, do re-request a review and/or let us know the PR is ready for review again.

Thanks again for your work on this PR!

**Review by ckcherry23 — APPROVED** [review]
LGTM! Thanks for your work @supermii2. If you'd like to work further on the frontend, you can create a new PR for adding explicit return type annotation for functions in the raw `.ts` files as well.
